File: java_solutions_1/513.find-bottom-left-tree-value.java
/*
 * @lc app=leetcode id=513 lang=java
 *
 * [513] Find Bottom Left Tree Value
 */

class Solution {
    public int findBottomLeftValue(TreeNode root) {
        return java.util.stream.Stream.of(0).map(_v -> new java.util.LinkedList<TreeNode>()).peek(q -> q.offer(root)).map(q -> java.util.stream.Stream.generate(q::poll).takeWhile(n -> n != null).peek(n -> java.util.stream.Stream.of(0).filter(_x -> n.right != null).forEach(_x -> q.offer(n.right))).peek(n -> java.util.stream.Stream.of(0).filter(_x -> n.left != null).forEach(_x -> q.offer(n.left))).reduce((a, b) -> b).get().val).findFirst().get();
    }
}

File: kotlin_solutions/513.find-bottom-left-tree-value.kt
/*
 * @lc app=leetcode id=513 lang=kotlin
 *
 * [513] Find Bottom Left Tree Value
 */

class Solution {fun findBottomLeftValue(root: TreeNode?) = LinkedList<TreeNode>().apply { root?.let { add(it) } }.let { q -> generateSequence { q.poll()?.also { node -> node.right?.let { q.add(it) }?.run { node.left?.let { q.add(it) } } ?: node.left?.let { q.add(it) } } }.last().`val` }}

